File: Week_02/G20190343020169/binary-tree-inorder-traversal.cpp
/*
 * @lc app=leetcode.cn id=94 lang=cpp
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include<vector>
#include<ctype.h>
#include<iostream>

using namespace std;

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;

        if(!root){
            return result;
        }

        vector<TreeNode*> v;
        v.push_back(root);
        while(v.size() > 0){
            TreeNode* pNode = v.back();
            if(pNode->left){
                v.push_back(pNode->left);
                pNode->left = NULL;
            }else{
                v.pop_back();
                result.push_back(pNode->val);
                if(pNode->right) v.push_back(pNode->right);
            }
        }   
        return result;  
    }

    vector<int> recursive_inorder(TreeNode* root){
        vector<int> v;
        inorderTraversal1(v,root);
        return v;
    }

    void inorderTraversal1(vector<int> &v,TreeNode* root){
        if(root == NULL){
            return;
        }
        inorderTraversal1(v, root->left);
        v.push_back(root->val);
        inorderTraversal1(v, root->right);
    }
};

File: Week_02/G20190343020023/binary-tree-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func inorderTraversal(root *TreeNode) []int {
    var result []int
    if root == nil {
        return result
    }
    return helper(root, result)
}

func helper(root *TreeNode, result []int) []int {
    if root.Left != nil {
        result = helper(root.Left, result)
    }
    result = append(result, root.Val)
    if root.Right != nil {
        result = helper(root.Right, result)
    }
    return result
}